src/main/kotlin/org/rust/cargo/toolchain/Cargo.kt

```diff
@@ -152,28 +150,39 @@ class Cargo(private val cargoExecutable: Path) {
             }
 
             if (settings.compileAllTargets && checkSupportForBuildCheckAllTargets()) add("--all-targets")
-            if (settings.useOffline) add("-Zoffline")
             addAll(ParametersListUtil.parse(settings.externalLinterArguments))
         }
 
         val useClippy = settings.externalLinter == ExternalLinter.CLIPPY
             && !checkNeedInstallClippy(project, cargoProjectDirectory)
         val checkCommand = if (useClippy) "clippy" else "check"
-        return CargoCommandLine(checkCommand, cargoProjectDirectory, arguments).execute(owner, ignoreExitCode = true)
+        return CargoCommandLine(checkCommand, cargoProjectDirectory, arguments)
+            .execute(project, owner, ignoreExitCode = true)
     }
 
-    fun toColoredCommandLine(commandLine: CargoCommandLine): GeneralCommandLine =
-        toGeneralCommandLine(commandLine, colors = true)
+    fun toColoredCommandLine(project: Project, commandLine: CargoCommandLine): GeneralCommandLine =
+        toGeneralCommandLine(project, commandLine, colors = true)
 
-    fun toGeneralCommandLine(commandLine: CargoCommandLine): GeneralCommandLine =
-        toGeneralCommandLine(commandLine, colors = false)
+    fun toGeneralCommandLine(project: Project, commandLine: CargoCommandLine): GeneralCommandLine =
+        toGeneralCommandLine(project, commandLine, colors = false)
 
-    private fun toGeneralCommandLine(commandLine: CargoCommandLine, colors: Boolean): GeneralCommandLine =
+    private fun toGeneralCommandLine(project: Project, commandLine: CargoCommandLine, colors: Boolean): GeneralCommandLine =
         with(patchArgs(commandLine, colors)) {
             val parameters = buildList<String> {
                 if (channel != RustChannel.DEFAULT) {
                     add("+$channel")
                 }
+                if (project.rustSettings.useOffline) {
+                    val cargoProject = findCargoProject(project, additionalArguments, workingDirectory)
+                    val rustcVersion = cargoProject?.rustcInfo?.version?.semver
+                    if (rustcVersion != null) {
+                        if (rustcVersion < RUST_1_36) {
+                            add("-Zoffline")
+                        } else {
+                            add("--offline")
+                        }
```

Comment:
It's just a matter of taste but I'd prefer to replace nested `if`s with single `when`
Something like the following
```kotlin
when {
    rustcVersion == null -> Unit
    rustcVersion < RUST_1_36 -> add("-Zoffline")
    else -> add("--offline")
}
```